dotCMS/src/main/java/com/dotcms/analytics/attributes/CustomAttributeAPIImpl.java

```diff
@@ -0,0 +1,169 @@
+
+        loadCache();
+    }
+
+    /**
+     * Loads all mappings into the cache from the underlying storage.
+     * Wraps any {@link DotDataException} in a {@link DotRuntimeException}.
+     */
+    private void loadCache() {
+        try {
+            customAttributeFactory.getAll()
+                    .forEach((key, value) -> customAttributeCache.put(key, value));
+        } catch (DotDataException e) {
+            throw new DotRuntimeException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @WrapInTransaction
+    @CloseDBIfOpened
+    public synchronized void checkCustomPayloadValidation(final String eventTypeName,
+                                             final Map<String, Object> customPayload)
+            throws MaxCustomAttributesReachedException, DotDataException {
+
+        Map<String, String> currentCustomAttributesMatch = getCustomAttributesMatchFromCache(eventTypeName);
+
+        final List<String> newCustomAttributes = getNewlyAttributes(customPayload, currentCustomAttributesMatch);
+
+        Map<String, String> customAttributesMatch = currentCustomAttributesMatch != null ?
+                new HashMap<>(currentCustomAttributesMatch) : new HashMap<>();
+
+        if (!newCustomAttributes.isEmpty()) {
+            assignMatchToNewAttributes(eventTypeName, customAttributesMatch, newCustomAttributes);
+        }
+
+        int currentTotalMatch = customAttributesMatch.size();
+
+        if (currentTotalMatch > MAX_LIMIT_CUSTOM_ATTRIBUTE_BY_EVENT) {
+            throw new MaxCustomAttributesReachedException(eventTypeName, MAX_LIMIT_CUSTOM_ATTRIBUTE_BY_EVENT);
+        }
+    }
+
+    /**
+     * Retrieves the attribute mapping for the provided event type from the cache, attempting to
+     * lazily load the cache if a miss occurs.
+     *
+     * @param eventTypeName the event type name.
+     * @return the mapping of attribute name to database column, or {@code null} if not found.
+     */
+    private Map<String, String> getCustomAttributesMatchFromCache(String eventTypeName) {
+        Map<String, String> customAttributesMatchFromCache = customAttributeCache.get(eventTypeName);
+
+        if (customAttributesMatchFromCache == null) {
+            loadCache();
+            customAttributesMatchFromCache = customAttributeCache.get(eventTypeName);
+        }
+        return customAttributesMatchFromCache;
+    }
+
+    /**
+     * Persists and caches new attribute mappings for the given event type.
+     */
+    private void assignMatchToNewAttributes(
+            final String eventTypeName,
+            final Map<String, String> customAttributesMatch,
+            final List<String> newCustomAttributes) throws DotDataException {
+
+        int index = customAttributesMatch.size() + 1;
+
+        for (final String newCustomAttribute : newCustomAttributes) {
+            customAttributesMatch.put(newCustomAttribute, CUSTOM_ATTRIBUTE_KEY + index);
```

Comment:
you are doing post increment, could be done in the same line such as   customAttributesMatch.put(newCustomAttribute, CUSTOM_ATTRIBUTE_KEY + index++);